### PR TITLE
Small refactorings for the `Merge` module

### PR DESCRIPTION
**Pull request by jorisdral**
See the commit messages for descriptions of the changes.

These changes are also in preparation for giving merges a `RefCounter`

### CODE REVIEW

File: src/Database/LSMTree/Internal/Merge.hs
Comment (jorisdral) [inline src/Database/LSMTree/Internal/Merge.hs:120]:
```suggestion
-- | Like 'steps', but calling 'complete' once the merge is finished.
```

File: src/Database/LSMTree/Internal/Merge.hs
Comment (jorisdral) [inline src/Database/LSMTree/Internal/Merge.hs:131]:
```suggestion
-- | Like 'steps', but calling 'complete' once the merge is finished.
```

File: src/Database/LSMTree/Internal/Merge.hs
Comment (jorisdral) [inline src/Database/LSMTree/Internal/Merge.hs:106]:
```suggestion
-- 'Readers', this function will throw an error.
```

File: src/Database/LSMTree/Internal/Merge.hs
Comment (mheinzel) [inline src/Database/LSMTree/Internal/Merge.hs:118]:
```suggestion
        Run.fromMutable mergeHasFS mergeHasBlockIO mergeCaching
                              (RefCount 1) mergeBuilder
```

**Review by mheinzel — APPROVED** [review]

File: src/Database/LSMTree/Internal/Merge.hs
Comment (mheinzel) [inline src/Database/LSMTree/Internal/Merge.hs:115]: Now that I think about it, we could make this API a bit safer by taking some `CompletedMerge` newtype argument that `steps` only returns in the case where the merge is complete.
Doesn't buy us a lot, unfortunately. We still need `mergeLastStepResult` for checking the opposite though (`steps` only being called on in-progress merges).

**Review by dcoutts — APPROVED** [review]
LGTM! I also endorse the review comments from @jorisdral and @mheinzel, but none extra myself.

File: src/Database/LSMTree/Internal/Merge.hs
Comment (dcoutts) [reply src/Database/LSMTree/Internal/Merge.hs:115]: Mmm, not clear to me either way.

Comment (jorisdral) [reply src/Database/LSMTree/Internal/Merge.hs:115]: Hmm yes, though I think it would complicate the `MergeSchedule` code a bit. With scheduling, `complete` is not necessarily called immediately after `steps`. Well, that's if we follow the prototype precisely, but we might deviate from that once we get to actually implementing it. For now, I'll leave the code as is

File: src/Database/LSMTree/Internal/Merge.hs
Comment (recursion-ninja) [inline src/Database/LSMTree/Internal/Merge.hs:91]: You could name the first two function arguments `mergeHasFS` and `mergeHasBlockIO` to omit the explicit record assignment. Your call.

**Review by recursion-ninja — COMMENTED** [review]
Overall, LGTM *(after the fact)*.

**Pull request merged**
